service/storage_proxy.cc

```diff
@@ -5123,12 +5132,22 @@ void storage_proxy::init_messaging_service(shared_ptr<migration_manager> mm) {
                 opts.request = da == query::digest_algorithm::none ? query::result_request::only_result : query::result_request::result_and_digest;
                 auto timeout = t ? *t : db::no_timeout;
                 return p->query_result_local(std::move(s), cmd, std::move(pr2.first), opts, trace_state_ptr, timeout);
-            }).finally([&trace_state_ptr, src_ip] () mutable {
+            }).then_wrapped([this, &trace_state_ptr, src_ip] (future<rpc::tuple<foreign_ptr<lw_shared_ptr<query::result>>, cache_temperature>> f) mutable {
+                using final_tuple_type = rpc::tuple<foreign_ptr<lw_shared_ptr<query::result>>, cache_temperature, std::optional<replica::exception_variant>>;
                 tracing::trace(trace_state_ptr, "read_data handling is done, sending a response to /{}", src_ip);
+                if (!f.failed()) {
+                    auto&& [result, temp] = f.get();
+                    return make_ready_future<final_tuple_type>(std::move(result), std::move(temp), std::nullopt);
+                } else if (features().cluster_supports_typed_errors_in_read_rpc()) {
```

Comment:
So far we usually check features on the coordinator side, not replica. The reasoning is that features aren't enabled simultaneously on all nodes. I don't think it's a problem here.

Reply:
(we usually have the coordinator enable a bit in the request, and the replica checks the request-specific bit)

@psarna can you think of a reason why checking on the replica would be problematic?

Reply:
Our features are a one-way ticket, once the whole cluster supports it, there's no way back. So it sounds perfectly fine to check the feature on a replica, because if it returns true, its coordinator should also be aware of it.